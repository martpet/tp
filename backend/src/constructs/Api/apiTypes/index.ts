export * from './EndpointsCookies';
export * from './EndpointsEnvironments';
export * from './EndpointsHeaders';
export * from './EndpointsOptions';
export * from './EndpointsQueryStrings';
export * from './OauthCookieProps';
export * from './OauthTokens';
