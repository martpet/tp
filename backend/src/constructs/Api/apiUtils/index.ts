export * from './cookieName';
export * from './errorResponse';
export * from './handlerWrapper';
export * from './itResolvesWithApiError';
