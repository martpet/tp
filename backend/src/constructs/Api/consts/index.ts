export * from './authorizationHeader';
export * from './defaultCachePolicyProps';
export * from './defaultSecurityHeadersBehavior';
export * from './lambdaEdgeViewerEvent';
