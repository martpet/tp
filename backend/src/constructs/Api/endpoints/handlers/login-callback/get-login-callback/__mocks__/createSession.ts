export const createSession = vi.fn().mockResolvedValue('dummySessionCookie');
