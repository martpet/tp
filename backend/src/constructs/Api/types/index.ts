export * from './EndpointCookies';
export * from './EndpointEnvVars';
export * from './EndpointHeaders';
export * from './EndpointsQueryStrings';
export * from './LambdaEdgeOriginRequestHandler';
export * from './LambdaEdgeResponse';
export * from './OauthCookieProps';
export * from './OauthTokens';
