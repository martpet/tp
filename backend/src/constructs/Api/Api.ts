import { HttpApi } from '@aws-cdk/aws-apigatewayv2-alpha';
import { Fn, NestedStack } from 'aws-cdk-lib';
import { Distribution } from 'aws-cdk-lib/aws-cloudfront';
import { HttpOrigin } from 'aws-cdk-lib/aws-cloudfront-origins';
import { ARecord, RecordTarget } from 'aws-cdk-lib/aws-route53';
import { CloudFrontTarget } from 'aws-cdk-lib/aws-route53-targets';
import { Construct } from 'constructs';

import { Auth, Photos, Tables, Zone } from '~/constructs';
import { apiSubdomain, appEnvs } from '~/consts';
import { getEnvName } from '~/utils';

import { createDistroBehaviors } from './endpoints/authEdgeFunction/createDistroBehaviors';
import { createRoutes } from './endpoints/createRoutes';

type Props = {
  zone: Zone;
  auth: Auth;
  tables: Tables;
  photos: Photos;
};

export class Api extends NestedStack {
  constructor(scope: Construct, { zone, tables, auth, photos }: Props) {
    super(scope, 'Api');

    const envName = getEnvName(this);
    const { appDomain } = appEnvs[envName];
    const apiDomain = `${apiSubdomain}.${appDomain}`;
    const api = new HttpApi(this, 'HttpApi');
    const origin = new HttpOrigin(Fn.select(1, Fn.split('://', api.apiEndpoint)));

    const distribution = new Distribution(this, 'Distro', {
      defaultBehavior: { origin },
      domainNames: [apiDomain],
      certificate: zone.certificate,
      enableLogging: true,
      logIncludesCookies: true,
    });

    createRoutes({ scope, api, auth, tables, photos });

    createDistroBehaviors({
      scope,
      api,
      auth,
      tables,
      distribution,
      origin,
    });

    new ARecord(this, 'Alias', {
      zone: zone.hostedZone,
      recordName: apiSubdomain,
      target: RecordTarget.fromAlias(new CloudFrontTarget(distribution)),
    });
  }
}
