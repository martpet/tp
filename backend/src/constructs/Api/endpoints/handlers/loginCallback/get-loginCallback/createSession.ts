import { DynamoDBClient } from '@aws-sdk/client-dynamodb';
import { DynamoDBDocumentClient, PutCommand } from '@aws-sdk/lib-dynamodb';
import cookie from 'cookie';
import crypto from 'crypto';
import millis from 'milliseconds';

import { OauthTokens } from '~/constructs/Api/types';
import { cookieName } from '~/constructs/Api/utils';
import { refreshTokenValidityInDays, sessionsTableOptions } from '~/consts';
import { SessionsTableItem } from '~/types';

const ddbClient = new DynamoDBClient({});
const ddbDocClient = DynamoDBDocumentClient.from(ddbClient);

type CreateSessionProps = {
  tokens: OauthTokens;
};

export const createSession = async ({ tokens }: CreateSessionProps) => {
  const { idTokenPayload, refreshToken, idToken } = tokens;
  const { envName } = globalLambdaProps;
  const sessionId = crypto.randomUUID();
  const created = Date.now();
  const refreshTokenExpires = created + millis.days(refreshTokenValidityInDays);

  const putCommandItem: SessionsTableItem = {
    id: sessionId,
    userId: idTokenPayload.sub,
    created,
    refreshToken,
    refreshTokenExpires,
    idToken,
  };

  const putCommand = new PutCommand({
    TableName: sessionsTableOptions.tableName,
    Item: putCommandItem,
  });

  await ddbDocClient.send(putCommand);

  const sessionCookie = cookie.serialize(cookieName('sessionId'), sessionId, {
    secure: true,
    httpOnly: true,
    sameSite: envName !== 'personal' || 'none',
    expires: new Date(refreshTokenExpires),
  });

  return sessionCookie;
};
