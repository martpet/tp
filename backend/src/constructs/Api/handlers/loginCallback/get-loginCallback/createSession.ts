import { DynamoDBClient } from '@aws-sdk/client-dynamodb';
import { DynamoDBDocumentClient, PutCommand } from '@aws-sdk/lib-dynamodb';
import cookie from 'cookie';
import { nanoid } from 'nanoid';

import { cookieName } from '~/constructs/Api/apiUtils';
import { refreshTokenValidityInDays, sessionsTableOptions } from '~/consts';
import { EnvName, SessionsTableItem } from '~/types';
import { daysToMilliseconds } from '~/utils';

import { OauthTokens } from '../../../apiTypes';

const ddbClient = new DynamoDBClient({});
const ddbDocumentClient = DynamoDBDocumentClient.from(ddbClient);

type CreateSessionProps = {
  tokens: OauthTokens;
  envName: string;
};

export const createSession = async ({ tokens, envName }: CreateSessionProps) => {
  const { idTokenPayload, accessToken, refreshToken } = tokens;

  const sessionId = nanoid();
  const created = Date.now();
  const expires = created + daysToMilliseconds(refreshTokenValidityInDays);

  const putCommandItem: SessionsTableItem = {
    id: sessionId,
    userId: idTokenPayload.sub,
    accessToken,
    refreshToken,
    created,
    expires,
  };

  const putCommand = new PutCommand({
    TableName: sessionsTableOptions.tableName,
    Item: putCommandItem,
  });

  await ddbDocumentClient.send(putCommand);

  const sessionCookie = cookie.serialize(cookieName('session'), sessionId, {
    secure: true,
    httpOnly: true,
    sameSite: (envName as EnvName) === 'personal' ? 'none' : true,
    expires: new Date(expires),
  });

  return sessionCookie;
};
