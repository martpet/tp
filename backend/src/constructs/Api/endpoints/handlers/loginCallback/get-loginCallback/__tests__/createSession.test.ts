import { DynamoDBDocumentClient, PutCommand } from '@aws-sdk/lib-dynamodb';
import { mockClient } from 'aws-sdk-client-mock';
import cookie from 'cookie';

import { OauthTokens } from '~/constructs/Api/types';

import { createSession } from '../createSession';

const ddbMock = mockClient(DynamoDBDocumentClient);

vi.mock('cookie');
vi.mock('nanoid');

const args = [
  {
    tokens: {
      accessToken: 'dummyAccessToken',
      refreshToken: 'dummyRefreshToken',
      idToken: 'dummyIdToken',
      idTokenPayload: { sub: 'dummySub ' },
    } as OauthTokens,
    envName: 'production',
  },
] as Parameters<typeof createSession>;

beforeEach(() => {
  ddbMock.reset();
});

describe('createSession', () => {
  it('calls "cookie.serialize" with correct args', async () => {
    await createSession(...args);
    expect(vi.mocked(cookie.serialize).mock.calls).toMatchSnapshot();
  });

  it('sends "PutCommand" to DynamoDB with correct args', async () => {
    await createSession(...args);
    expect(ddbMock.commandCalls(PutCommand)[0].args[0].input).toMatchSnapshot();
  });

  it('resolves with a correct value', () => {
    return expect(createSession(...args)).resolves.toMatchSnapshot();
  });

  describe('when "envName" is "personal"', () => {
    const argsClone = structuredClone(args);
    argsClone[0].envName = 'personal';

    it('calls "cookie.serialize" with correct args', async () => {
      await createSession(...argsClone);
      expect(vi.mocked(cookie.serialize).mock.calls).toMatchSnapshot();
    });
  });
});
