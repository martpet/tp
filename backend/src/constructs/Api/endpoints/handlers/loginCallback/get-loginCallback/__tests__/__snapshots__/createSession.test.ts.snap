// Vitest Snapshot v1

exports[`createSession > calls "cookie.serialize" with correct args 1`] = `
[
  [
    "sessionId",
    "dummyNanoid",
    {
      "expires": 2023-01-01T00:00:00.000Z,
      "httpOnly": true,
      "sameSite": true,
      "secure": true,
    },
  ],
]
`;

exports[`createSession > resolves with a correct value 1`] = `"dummyCookie"`;

exports[`createSession > sends "PutCommand" to DynamoDB with correct args 1`] = `
{
  "Item": {
    "created": 1640995200000,
    "id": "dummyNanoid",
    "idToken": "dummyIdToken",
    "refreshToken": "dummyRefreshToken",
    "refreshTokenExpires": 1672531200000,
    "userId": "dummySub ",
  },
  "TableName": "TripPics-Sessions",
}
`;

exports[`createSession > when "globalLambdaProps.cdkNev" is "personal" > calls "cookie.serialize" with correct args 1`] = `
[
  [
    "sessionId",
    "dummyNanoid",
    {
      "expires": 2023-01-01T00:00:00.000Z,
      "httpOnly": true,
      "sameSite": "none",
      "secure": true,
    },
  ],
]
`;
